### PR TITLE
python312Packages.distutils: init at 0.70.0

### LINKED ISSUE #326296: Build failure: python312Packages.python-ldap
### Steps To Reproduce

1. build python312Packages.python-ldap

### Build log

```
writing byte-compilation script '/build/tmpz8pb47i7.py'
/nix/store/z7xxy35k7620hs6fn6la5fg2lgklv72l-python3-3.12.4/bin/python3.12 /build/tmpz8pb47i7.py
Traceback (most recent call last):
  File "/build/tmpz8pb47i7.py", line 1, in <module>
    from distutils.util import byte_compile
ModuleNotFoundError: No module named 'distutils'
error: command '/nix/store/z7xxy35k7620hs6fn6la5fg2lgklv72l-python3-3.12.4/bin/python3.12' failed with exit code 1

ERROR Backend subprocess exited when trying to invoke build_wheel
```

### Additional context
This breaks `sssd`, `autofs`...

### Notify maintainers

None :/

### Metadata

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.6.37, NixOS, 24.11 (Vicuna), 24.11.20240708.655a58a`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.4`
 - nixpkgs: `/nix/store/yzkx7gx2cn73dh3c7zg5rvgkp3yli1cd-source`
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### LINKED ISSUE #326920: python3.11-distorm3 - error during building
### Describe the bug
During the installation of Python3.11 Distorm3 package, I get:
```
============================= test session starts ==============================
platform linux -- Python 3.11.9, pytest-8.2.2, pluggy-1.5.0
rootdir: /build/source
go: downloading google.golang.org/protobuf v1.33.0
go: downloading github.com/go-git/gcfg v1.5.1-0.20230307220236-3a3c6141e376
collected 366 items / 3 errors                                                 

==================================== ERRORS ====================================
___________________ ERROR collecting python/test_distorm3.py ___________________
/nix/store/lqiim8h8978ysvnv4gwajxdmgyy3bxxl-python3.11-pytest-8.2.2/lib/python3.11/site-packages/_pytest/runner.py:341: in from_call
    result: Optional[TResult] = func()
/nix/store/lqiim8h8978ysvnv4gwajxdmgyy3bxxl-python3.11-pytest-8.2.2/lib/python3.11/site-packages/_pytest/runner.py:389: in collect
    return list(collector.collect())
/nix/store/lqiim8h8978ysvnv4gwajxdmgyy3bxxl-python3.11-pytest-8.2.2/lib/python3.11/site-packages/_pytest/unittest.py:90: in collect
    self.session._fixturemanager.parsefactories(self.newinstance(), self.nodeid)
/nix/store/lqiim8h8978ysvnv4gwajxdmgyy3bxxl-python3.11-pytest-8.2.2/lib/python3.11/site-packages/_pytest/unittest.py:75: in newinstance
    return self.obj("runTest")
E   TypeError: Test.__init__() takes 1 positional argument but 2 were given
___________________ ERROR collecting python/test_distorm3.py ___________________
/nix/store/lqiim8h8978ysvnv4gwajxdmgyy3bxxl-python3.11-pytest-8.2.2/lib/python3.11/site-packages/_pytest/runner.py:341: in from_call
    result: Optional[TResult] = func()
/nix/store/lqiim8h8978ysvnv4gwajxdmgyy3bxxl-python3.11-pytest-8.2.2/lib/python3.11/site-packages/_pytest/runner.py:389: in collect
    return list(collector.collect())
/nix/store/lqiim8h8978ysvnv4gwajxdmgyy3bxxl-python3.11-pytest-8.2.2/lib/python3.11/site-packages/_pytest/unittest.py:90: in collect
    self.session._fixturemanager.parsefactories(self.newinstance(), self.nodeid)
/nix/store/lqiim8h8978ysvnv4gwajxdmgyy3bxxl-python3.11-pytest-8.2.2/lib/python3.11/site-packages/_pytest/unittest.py:75: in newinstance
    return self.obj("runTest")
E   TypeError: InstBin.__init__() missing 3 required positional arguments: 'mode', 'features', and 'address'
___________________ ERROR collecting python/test_distorm3.py ___________________
/nix/store/lqiim8h8978ysvnv4gwajxdmgyy3bxxl-python3.11-pytest-8.2.2/lib/python3.11/site-packages/_pytest/runner.py:341: in from_call
    result: Optional[TResult] = func()
/nix/store/lqiim8h8978ysvnv4gwajxdmgyy3bxxl-python3.11-pytest-8.2.2/lib/python3.11/site-packages/_pytest/runner.py:389: in collect
    return list(collector.collect())
/nix/store/lqiim8h8978ysvnv4gwajxdmgyy3bxxl-python3.11-pytest-8.2.2/lib/python3.11/site-packages/_pytest/unittest.py:90: in collect
    self.session._fixturemanager.parsefactories(self.newinstance(), self.nodeid)
/nix/store/lqiim8h8978ysvnv4gwajxdmgyy3bxxl-python3.11-pytest-8.2.2/lib/python3.11/site-packages/_pytest/unittest.py:75: in newinstance
    return self.obj("runTest")
E   TypeError: Inst.__init__() missing 3 required positional arguments: 'mode', 'instNo', and 'features'
=========================== short test summary info ============================
ERROR python/test_distorm3.py::Test - TypeError: Test.__init__() takes 1 positional argument but 2 were given
ERROR python/test_distorm3.py::InstBin - TypeError: InstBin.__init__() missing 3 required positional arguments: 'mod...
ERROR python/test_distorm3.py::Inst - TypeError: Inst.__init__() missing 3 required positional arguments: 'mode',...
!!!!!!!!!!!!!!!!!!! Interrupted: 3 errors during collection !!!!!!!!!!!!!!!!!!!!
============================== 3 errors in 29.05s ==============================
/nix/store/5r0df66ikad3xw06azlqvswcvncll8wa-stdenv-linux/setup: line 1641: pop_var_context: head of shell_variables not a function context
go: downloading github.com/pjbgf/sha1cd v0.3.0
error: builder for '/nix/store/0mcd4xx6605j7fagjwqvafi03q9ni4sl-python3.11-distorm3-3.5.2.drv' failed with exit code 2
```

### Notify maintainers

@fabaff 
### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.6.37, Athena OS, 24.11 (Vicuna), 24.11pre652179.7e7c39ea35c5`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.5`
 - channels(root): `"home-manager, nixos, nixpkgs"`
 - nixpkgs: `/home/athena/.nix-defexpr/channels/nixpkgs`
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### LINKED ISSUE #328346: Build failure: python312Packages.distorm3
### Steps To Reproduce

Steps to reproduce the behavior (replace `/dev/sda` with the disk you're trying to install NixOS onto):
1. `fdisk /dev/sda`
2. `mkfs.vfat -F32 /dev/sda1`
3. `mkfs.bcachefs /dev/sda2`
4. `mount -t bcachefs /dev/sda2 /mnt`
5. `mkdir /mnt/boot`
6. `mount -t vfat /dev/sda1 /mnt/boot`
7. `nixos-generate-config --root /mnt`
8. `wget -O /mnt/etc/nixos/configuration.nix https://gist.github.com/kennystrawnmusic/05e6bcea71e2d08f2414b9681b2b0de8/raw/ce55548890583f5b498404d8ba7cc55768ed369a/configuration.nix`
9. `nixos-install -I nixpkgs=https://github.com/NixOS/nixpkgs/archive/staging.tar.gz -I nixos=https://github.com/NixOS/nixpkgs/archive/staging.tar.gz`

### Build log

```
building the configuration in /mnt/etc/nixos/configuration.nix...
error:
       … while calling the 'head' builtin
         at /mnt/nix/store/g6k7p7d1480l6d307sqbcklfszva6g1s-source/lib/attrsets.nix:1575:11:
         1574|         || pred here (elemAt values 1) (head values) then
         1575|           head values
             |           ^
         1576|         else

       … while evaluating the attribute 'value'
         at /mnt/nix/store/g6k7p7d1480l6d307sqbcklfszva6g1s-source/lib/modules.nix:821:9:
          820|     in warnDeprecation opt //
          821|       { value = addErrorContext "while evaluating the option `${showOption loc}':" value;
             |         ^
          822|         inherit (res.defsFinal') highestPrio;

       … while evaluating the option `system.build.toplevel':

       … while evaluating definitions from `/mnt/nix/store/g6k7p7d1480l6d307sqbcklfszva6g1s-source/nixos/modules/system/activation/top-level.nix':

       … while evaluating the option `system.systemBuilderArgs':

       … while evaluating definitions from `/mnt/nix/store/g6k7p7d1480l6d307sqbcklfszva6g1s-source/nixos/modules/system/activation/activatable-system.nix':

       … while evaluating the option `system.activationScripts.etc.text':

       … while evaluating definitions from `/mnt/nix/store/g6k7p7d1480l6d307sqbcklfszva6g1s-source/nixos/modules/system/etc/etc-activation.nix':

       … while evaluating definitions from `/mnt/nix/store/g6k7p7d1480l6d307sqbcklfszva6g1s-source/nixos/modules/system/etc/etc.nix':

       … while evaluating the option `environment.etc.dbus-1.source':

       (stack trace truncated; use '--show-trace' to show the full, detailed trace)

       error: distorm3-3.5.2 not supported for interpreter python3.12

```

### Additional context

Actually used staging tarballs to test the supposed #325968 fix; still having problems despite that

### Notify maintainers

@fabaff 

### Metadata

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
 - system: `"x86_64-linux"`
 - host os: `Linux 6.9.9-zen1-1-zen, Arch Linux, noversion, rolling`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.23.3`
 - channels(root): `"nixpkgs"`
 - nixpkgs: `/home/realkstrawn93/.nix-defexpr/channels/nixpkgs`
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by dotlambda**
## Description of changes
fixes https://github.com/NixOS/nixpkgs/issues/328346
fixes https://github.com/NixOS/nixpkgs/issues/326296
closes https://github.com/NixOS/nixpkgs/pull/326321
fixes https://github.com/NixOS/nixpkgs/issues/326920

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by emilazy** [comment]
Would this same issue also be solved by fixing the issue that causes `setuptools` to [not provide `distutils` for us](https://github.com/NixOS/nixpkgs/pull/326321#discussion_r1675743173)? I would personally feel more comfortable with that as an interim solution than depending on another random third party. There’s no commit history in this repository either, and [the `distutils` licence mentioning `nntplib`](https://github.com/youknowone/python-deadlib/blob/a67a7c6338a08e9da2ffae268c3cab3be02d5f83/distutils/LICENSE) doesn’t fill me with confidence.

File: pkgs/development/python-modules/standard-distutils/default.nix [hunk 1]
```diff
@@ -0,0 +1,32 @@
+{
+  lib,
+  buildPythonPackage,
+  fetchPypi,
+  setuptools,
+}:
+
+buildPythonPackage rec {
+  pname = "standard-distutils";
```
Comment (kfollesdal) [inline pkgs/development/python-modules/standard-distutils/default.nix:9]: I would be more comfortable to use a more official source https://github.com/pypa/distutils, that is the source for the `distutils` that `setuptools` offer.

**Comment by dotlambda** [comment]
> Would this same issue also be solved by fixing the issue that causes `setuptools` to [not provide `distutils` for us](https://github.com/NixOS/nixpkgs/pull/326321#discussion_r1675743173)?

Yes.

> I would personally feel more comfortable with that as an interim solution than depending on another random third party.

Same. But that's a mass rebuild.

**Comment by emilazy** [comment]
What if we add a `distutils` package that just re‐exports the version from our `setuptools` package at the right path? That avoids a mass rebuild and offers a simple path towards migrating it back into `setuptools` package during a staging cycle (but it might actually be valuable for us to know what is using `distutils`, since ideally nothing should be).

**Comment by dotlambda** [comment]
> What if we add a `distutils` package that just re‐exports the version from our `setuptools` package at the right path?

It seemed easier to use https://github.com/pypa/distutils because that comes with `pyproject.toml`. So that's what I did.

**Comment by kfollesdal** [comment]
I have made a proposal to fix `setuptools` support for `distutils` so it behaves as expected. I prefer to fix `setuptools` rather than introducing an extra package. #328415

**Comment by dotlambda** [comment]
I'll merge this now cause the other option needs a whole staging cycle. If we agree the other solution is better we can revert this on staging.

**Pull request merged**